File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithComponents.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

@KoraApp
public interface AppWithComponents {
    @Root
    default Class1 class1(Class2 class2) {
        return new Class1(class2);
    }

    @Root
    default Class2 class2(Interface1 class3) {
        return new Class2(class3);
    }

    @Root
    default Class3 class3() {
        return new Class3();
    }

    @Root
    default GenericClass<? extends Class3> genericExtends(Class3 class3) {
        return new GenericClass<>(class3);
    }

    @Root
    default GenericClass<? super Class3> genericSuper(Class3 class3) {
        return new GenericClass<>(class3);
    }


    record Class1(Class2 class2) {}

    record Class2(Interface1 class3) {}

    interface Interface1 {}

    class Class3 implements Interface1 {}

    record GenericClass<T>(T t) {}

}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithComponents.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root

@KoraApp
interface AppWithComponents {
    @Root
    fun class1(class2: Class2) = Class1(class2)

    @Root
    fun class2(class3: Interface1) = Class2(class3)

    fun class3() = Class3()

    @Root
    fun genericExtends(class3: Class3) = GenericClass(class3)

    @Root
    fun genericSuper(class3: Class3) = GenericClass(class3)

    data class Class1(val class2: Class2)

    data class Class2(val class3: Interface1)

    interface Interface1
    class Class3 : Interface1
    data class GenericClass<T>(val t: T)
}
